hubblestack/extmods/modules/pulsar.py

```diff
@@ -66,17 +66,529 @@ def _enqueue(revent):
     '''
     __context__['pulsar.queue'].append(revent)
 
+class ConfigManager(object):
+    _config = {}
+    _last_update = 0
+
+    @property
+    def config(self):
+        if self.stale():
+            self.update()
+        return self.nc_config
+
+    @property
+    def nc_config(self):
+        return self.__class__._config
+
+    @nc_config.setter
+    def nc_config(self, v):
+        self.__class__._config = v
+
+    @config.setter
+    def config(self, v):
+        return self.nc_config.update(v)
+
+    @property
+    def last_update(self):
+        return self.__class__._last_update
+
+    @last_update.setter
+    def last_update(self, v):
+        self.__class__._last_update = v
+
+    def freshness(self, freshness_limit=2):
+        t = time.time()
+        return (t - self.last_update <= freshness_limit)
+
+    def stale(self):
+        if (time.time() - self.last_update) >= self.nc_config.get('refresh_interval', 300):
+            return True
+        return False
+
+    def format_path(self, path):
+        path  = os.path.abspath(path)
+        fname = os.path.basename(path)
+        cpath = self.path_of_config(path)
+        dname = path if os.path.isdir(path) else os.path.dirname(path)
+        return cpath, path, dname, fname
+
+    def path_config(self, path, falsifyable=False):
+        config = self.nc_config
+        if falsifyable and path not in config:
+            return False
+        c = collections.defaultdict(lambda: False)
+        c.update( config.get(path, {}) )
+        return c
+
+    def path_of_config(self, path):
+        ncc = self.nc_config
+        while len(path)>1 and path not in ncc:
+            path = os.path.dirname(path)
+        return path
+
+    def _abspathify(self):
+        c = self.nc_config
+        for k in tuple(c):
+            if k.startswith('/'):
+                l = os.path.abspath(k)
+                if k != l:
+                    c[l] = c.pop(k)
+
+    def update(self):
+        config = self.nc_config
+        to_set = __opts__.get('pulsar', {})
+
+        # Is there a better way to tell if __opts__ updated?
+        # Is it worth checking anyway? Seems only to come up in tests/
+        # todo?: attempt to re-read /etc/hubble/hubble sometimes?
+        counter = len( set(config).symmetric_difference( set(to_set) ) )
+        if counter == 0:
+            for k in config:
+                if config[k] != to_set[k]:
+                    counter += 1
```

Comment:
We have no plans to reload __opts__ at this time. I would skip this to save processing time.

Reply:
I doubt it adds much processing time. ... I'll go ahead and remove it though.